### DISCUSSION TITLE
[@types/leaflet-polylinedecorator] definitions are not correct

### DISCUSSION THREAD

**Discussion by lajerca** [DefinitelyTyped/DefinitelyTyped#61478] [Issues with a @types package]
the types of leaflet-polylinedecorator don't work for me out of the box. And I couldn't find a way to import or extend the leaflet namespace with these typings. I always get an error like `Property 'polylineDecorator' does not exist on type ...`

What works though is wrapping the types definition like this: 
```typescript
declare global {
  namespace L {
      namespace Symbol {
        // ....stuff
      };

      function polylineDecorator();

      // and so on....
  }
}
```
instead of declaring a module "leaflet" as it is here
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7b179c7a9a4aa4ff13f0608606ae10a94349014/types/leaflet-polylinedecorator/index.d.ts#L11

since in leaflet definitions the namespace is defined like this:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7b179c7a9a4aa4ff13f0608606ae10a94349014/types/leaflet/index.d.ts#L13

[a comment by an automated account is not shown]

**Comment by BePo65** [comment]
What you are doing is extending the global `L` and not the module `leaflet`.

Where do you use leaflet (e.g. angular, react)?
What `import` line do you use?

In my program I use this type definition for a long while. Here is a code fragment of the imports:
```
: 
import {
  Layer,
  LayerGroup,
  map, Map,
  MapOptions,
  polyline, Polyline,
  polylineDecorator, PolylineDecorator,
} from 'leaflet';
import 'leaflet-loading';
import 'leaflet-polylinedecorator';
import 'leaflet.icon.glyph';
import 'leaflet-easybutton';
import 'leaflet.fullscreen';
:
```

**Reply by lajerca** [reply]
I use it in a react project, but it shouldn't make a difference how or where I use it as long as type definitions are extended correctly.
I am surprised you say it works for you, since after many tests in my local environment I can't make it work.
The only way I can think of is using it outside of a module, which I never do.

### After some research and tests I found the answer:

The problem here is twofold:
1. using import statement in ambient declarations
2. mixing module and namespace declarations

What you say would work if leaflet was also exported as a global module and not as a namespace (tested locally, it would work). Modules become scoped as soon as a global import statement is added  like in this case. 
As stated in typescript docs:
https://www.typescriptlang.org/docs/handbook/modules.html
> In TypeScript, just as in ECMAScript 2015, any file containing a top-level import or export is considered a module. Conversely, a file without any top-level import or export declarations is treated as a script whose contents are available in the global scope (and therefore to modules as well).

so you might think that the culprit is this line, and removing it would just work. However that would result in leaflet module being overridden by and not extended by leaflet-polylinedecorator type.
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7b179c7a9a4aa4ff13f0608606ae10a94349014/types/leaflet-polylinedecorator/index.d.ts#L9

Since leaflet is exported as a namespace, the only option left I can think of is to adapt `@types/leaflet-polylinedecorator` to `@types/leaflet` by using the namespace pattern as I explained in my initial comment.

**Comment by BePo65** [comment]
As it is working for me here more information about my environment:
* I installed the following packages / type definitions:  
```
 "dependencies": {
    "leaflet": "^1.7.1",
    "leaflet-polylinedecorator": "^1.6.0",
 },
 "dependencies": {
    "@types/leaflet": "^1.7.9",
    "@types/leaflet-polylinedecorator": "^1.6.1",
 },
```
* I use angular 12;  
the tsconfig `moduleResolution` is `node` (i.e. 'commonjs').  
So can it be that your react project uses 'node12' or 'nodenext' (i.e. 'ECMAScript modules') in its tsconfig?

**Comment by lajerca** [comment]
in tsconfig.json I am using:
```
  "module": "CommonJS",
  "moduleResolution": "node",
```
but playing with other values doesn't work. My only option at the moment is to use the library without types

**Comment by BePo65** [comment]
very strange; perhaps I can create a little test project for playing around.

**Comment by BePo65** [comment]
As I said, I am not a react guy, so I looked around for react and leaflet and all I found was 'react leaflet'. And in the documentation and stackoverflow it looks like you cannot use native leaflet plugins in react leaflet.

This type definition is for typescript + the original leaflet package + the original polylinedecorator plugin and can be used e.g. in angular (what I do).

What framework / environment do you use. Are you sure that this type definition is for your environment?

**Reply by lajerca** [reply]
react is just another library like all the others in package.json. It shouldn't make a difference, neither should react-leaflet play any role in this. I am using multiple plugins in my projects with no issues.

It turned out there was something wrong in my environment, for some reason deleting node_modules and reinstalling fixed the issue.
Thanks for your help anyway and sorry for the waste of time.
